### PR TITLE
Basic types, classes and structs

### PR DESCRIPTION
**Pull request by stevehoag**
Split Types topic and added C# 7 literals features

### CODE REVIEW

**Comment by BillWagner** [comment]
Cool! I'll give this a review this evening.

**Comment by stevehoag** [comment]
I'm going to need to do another merge - I still had the Structs topic open so the latest changes didn't get in.

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some changed files links of this time:
- Docset 1 - **core-docs**  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134)  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134)  

**Status**: Succeeded.  
[Open Publishing Report](https://opbuildstorageprod.blob.core.windows.net/report/2016%5C10%5C11%5C2a032abb-a76c-bfc8-4231-71ab612989a2%5CPullRequest%5C201610112116310438-1134%5Cworkflow_report.html).

File: docs/csharp/basic-types.md [hunk 1]
```diff
@@ -4,24 +4,80 @@ description: Learn about the core types (numerics, strings, and object) in all C
 keywords: .NET, .NET Core, C#
 author:  dotnet-bot
```
Comment (BillWagner) [inline docs/csharp/basic-types.md]: You should change this to you.

File: docs/csharp/classes.md [hunk 2]
```diff
@@ -4,24 +4,78 @@ description: Learn about the class types and how you create them
 keywords: .NET, .NET Core, C#
 author:  dotnet-bot
```
Comment (BillWagner) [inline docs/csharp/classes.md]: you should change this to you.

File: docs/csharp/classes.md [hunk 3]
```diff
@@ -4,24 +4,78 @@ description: Learn about the class types and how you create them
...
+  
+Reference types fully support *inheritance*, a fundamental characteristic of object-oriented programming. When you create a class, you can inherit from any other interface or class that is not defined as [sealed](https://msdn.microsoft.com/en-us/library/88c54tsw.aspx), and other classes can inherit from your class and override your virtual methods. For more information, see [Inheritance](https://msdn.microsoft.com/en-us/library/ms173149.aspx).
+
+## Declaring classes  
+Classes are declared by using the [class](https://msdn.microsoft.com/en-us/library/0b0thckt.aspx) keyword, as shown in the following example:  
+  
+[!code-csharp[Declaring Classes](../../samples/snippets/csharp/concepts/classes/declaring-classes.cs)]  
+  
+The **class** keyword is preceded by the access level. Because [public](https://msdn.microsoft.com/en-us/library/yzh058ae.aspx) is used in this case, anyone can create objects from this class. The name of the class follows the **class** keyword. The remainder of the definition is the class body, where the behavior and data are defined. Fields, properties, methods, and events on a class are collectively referred to as *class members*.
```
Comment (BillWagner) [inline docs/csharp/classes.md]: access level should be "access modifier".

File: docs/csharp/classes.md [hunk 4]
```diff
@@ -4,24 +4,78 @@ description: Learn about the class types and how you create them
...
+
+## Declaring classes  
+Classes are declared by using the [class](https://msdn.microsoft.com/en-us/library/0b0thckt.aspx) keyword, as shown in the following example:  
+  
+[!code-csharp[Declaring Classes](../../samples/snippets/csharp/concepts/classes/declaring-classes.cs)]  
+  
+The **class** keyword is preceded by the access level. Because [public](https://msdn.microsoft.com/en-us/library/yzh058ae.aspx) is used in this case, anyone can create objects from this class. The name of the class follows the **class** keyword. The remainder of the definition is the class body, where the behavior and data are defined. Fields, properties, methods, and events on a class are collectively referred to as *class members*.  
+  
+## Creating objects  
+Although they are sometimes used interchangeably, a class and an object are different things. A class defines a type of object, but it is not an object itself. An object is a concrete entity based on a class, and is sometimes referred to as an instance of a class.
```
Comment (BillWagner) [inline docs/csharp/classes.md]: I would drop this first clause. class and object should not be used interchangeably.

File: docs/csharp/structs.md [hunk 5]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
->
-> We welcome your input to help shape the scope and approach. You can track the status and provide input on this
-> [issue](https://github.com/dotnet/core-docs/issues/966) at GitHub.
-> 
-> If you would like to review early drafts and outlines of this topic, please leave a note with your contact information in the issue.
->
-> Learn more about how you can contribute on [GitHub](https://github.com/dotnet/core-docs/blob/master/CONTRIBUTING.md).
->
+# Structs
+A struct is a value type. When a struct is created, the variable to which the struct is assigned holds the struct's actual data. When the struct is assigned to a new variable, it is copied. The new variable and the original variable therefore contain two separate copies of the same data. Changes made to one copy do not affect the other copy.
```
Comment (BillWagner) [inline docs/csharp/structs.md]: Should `struct` be in inline style?

File: docs/csharp/structs.md [hunk 6]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
 ms.assetid: a7094b8c-7229-4b6f-82fc-824d0ea0ec40
```
Comment (BillWagner) [inline docs/csharp/structs.md]: author and date should change.

File: docs/csharp/structs.md [hunk 7]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
-> [issue](https://github.com/dotnet/core-docs/issues/966) at GitHub.
-> 
-> If you would like to review early drafts and outlines of this topic, please leave a note with your contact information in the issue.
->
-> Learn more about how you can contribute on [GitHub](https://github.com/dotnet/core-docs/blob/master/CONTRIBUTING.md).
->
+# Structs
+A struct is a value type. When a struct is created, the variable to which the struct is assigned holds the struct's actual data. When the struct is assigned to a new variable, it is copied. The new variable and the original variable therefore contain two separate copies of the same data. Changes made to one copy do not affect the other copy.
+
+Value types derive from @System.ValueType, which derives from @System.Object. Types that derive from @System.ValueType have special behavior in the CLR. Value type variables directly contain their values, which means that the memory is allocated inline in whatever context the variable is declared. There is no separate heap allocation or garbage collection overhead for value-type variables.
```
Comment (BillWagner) [inline docs/csharp/structs.md]: This is misleading, in a way. It reads like you must define structs by deriving from System.ValueType. That's all implicit.

File: docs/csharp/structs.md [hunk 8]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+  
+-   Unlike classes, structs can be instantiated without using a **new** operator.  
+  
+-   Structs can declare constructors that have parameters.  
+  
+-   A struct cannot inherit from another struct or class, and it cannot be the base of a class. All structs inherit directly from **System.ValueType**, which inherits from **System.Object**.  
+  
+-   A struct can implement interfaces.  
+  
+-   A struct can be used as a nullable type and can be assigned a null value.
```
Comment (BillWagner) [inline docs/csharp/structs.md]: A nullable corresponding type can be built from any struct type. A struct can't b used as a nullable type.

File: docs/csharp/structs.md [hunk 9]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+
+## Literal values  
+In C#, literal values receive a type from the compiler. You can specify how a numeric literal should be typed by appending a letter to the end of the number. For example, to specify that the value 4.56 should be treated as a float, append an "f" or "F" after the number: `4.56f`. If no letter is appended, the compiler will infer a type for the literal. For more information about which types can be specified with letter suffixes, see the reference pages for individual types in [Value Types](https://msdn.microsoft.com/en-us/library/s1ax56ch.aspx).
```
Comment (BillWagner) [inline docs/csharp/structs.md]: We should probably say "the compiler will infer the `double` type for the literal."

File: samples/snippets/csharp/concepts/structs/literals.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+Console.WriteLine(type);
+
+var x = 123_456;
+string s2 = "I can use an underscore as a digit separator: " + x;
+// Outputs: "I can use an underscore as a digit separator: 
+Console.WriteLine(s2);
+
+var b = 0b1010_1011_1100_1110_1111;
+string s3 = "I can specify bit patterns: " + b.ToString();
+// Outputs: "I can specify bit patterns:
```
Comment (BillWagner) [inline samples/snippets/csharp/concepts/structs/literals.cs]: we should add the value here.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Looks good. I made a few comments.

File: docs/csharp/basic-types.md [hunk 11]
```diff
@@ -4,24 +4,80 @@ description: Learn about the core types (numerics, strings, and object) in all C
...
+
+## Implicit Types, and Anonymous Types  
+As stated previously, you can implicitly type a local variable (but not class members) by using the [var](https://msdn.microsoft.com/en-us/library/bb383973.aspx) keyword. The variable still receives a type at compile time, but the type is provided by the compiler. For more information, see [Implicitly typed local variables](https://msdn.microsoft.com/en-us/library/bb384061.aspx).  
+  
+In some cases, it is inconvenient to create a named type for simple sets of related values that you do not intend to store or pass outside method boundaries. You can create *anonymous types* for this purpose. For more information, see [Anonymous types](https://msdn.microsoft.com/en-us/library/bb397696.aspx).
+
```
Comment (stevehoag) [inline docs/csharp/basic-types.md]: Good suggestion - done.

File: docs/csharp/classes.md (hunk 2 above)
Comment (stevehoag) [inline docs/csharp/classes.md]: Done

File: docs/csharp/classes.md (hunk 4 above)
Comment (stevehoag) [inline docs/csharp/classes.md]: Agreed - I removed it.

File: docs/csharp/structs.md (hunk 5 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

File: docs/csharp/structs.md (hunk 6 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Done.

File: docs/csharp/structs.md (hunk 7 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

File: docs/csharp/structs.md (hunk 8 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

File: docs/csharp/basic-types.md [hunk 12]
```diff
@@ -4,24 +4,80 @@ description: Learn about the core types (numerics, strings, and object) in all C
...
+  
+The types of method parameters and return values are specified in the method signature. The following signature shows a method that requires an [int](https://msdn.microsoft.com/en-us/library/5kzh1b5w.aspx) as an input argument and returns a string:  
+  
+[!code-csharp[Method Signature](../../samples/snippets/csharp/concepts/basic-types/method-signature.cs)]  
+  
+After a variable is declared, it cannot be re-declared with a new type, and it cannot be assigned a value that is not compatible with its declared type. For example, you cannot declare an [int](https://msdn.microsoft.com/en-us/library/5kzh1b5w.aspx) and then assign it a Boolean value of [true](https://msdn.microsoft.com/en-us/library/06d3w013.aspx). However, values can be converted to other types, for example when they are assigned to new variables or passed as method arguments. A *type conversion* that does not cause data loss is performed automatically by the compiler. A conversion that might cause data loss requires a *cast* in the source code. 
+
+For more information, see [Casting and type conversions](https://msdn.microsoft.com/en-us/library/ms173105.aspx)
```
Comment (mairaw) [inline docs/csharp/basic-types.md]: nit: missing punctuation in the end

File: docs/csharp/basic-types.md [hunk 13]
```diff
@@ -4,24 +4,80 @@ description: Learn about the core types (numerics, strings, and object) in all C
...
+  
+-   Each type in the CTS is defined as either a *value type* or a *reference type*. This includes all custom types in the .NET Framework class library and also your own user-defined types. Types that you define by using the [struct](https://msdn.microsoft.com/en-us/library/ah19swz4.aspx) keyword are value types; all the built-in numeric types are **structs**. For more information about value types, see [Structs](structs.md).Types that you define by using the [class](https://msdn.microsoft.com/en-us/library/0b0thckt.aspx) keyword are reference types. For more information about reference types, see [Classes](classes.md). Reference types and value types have different compile-time rules, and different run-time behavior.
```
Comment (mairaw) [inline docs/csharp/basic-types.md]: nit: Missing space before the sentence "Types that you define"

File: docs/csharp/classes.md [hunk 14]
```diff
@@ -4,24 +4,78 @@ description: Learn about the class types and how you create them
...
+
+[!code-csharp[Reference Types](../../samples/snippets/csharp/concepts/classes/reference-type.cs)]
+  
+When the object is created, the memory is allocated on the managed heap, and the variable holds only a reference to the location of the object. Types on the managed heap require overhead both when they are allocated and when they are reclaimed by the automatic memory management functionality of the CLR, which is known as *garbage collection*. However, garbage collection is also highly optimized, and in most scenarios it does not create a performance issue. For more information about garbage collection, see [Automatic memory management and garbage collection](../Standard/garbagecollection/gc.md).
```
Comment (mairaw) [inline docs/csharp/classes.md]: nit: missing comma in most scenarios

also: lowercase link (../standard/....)

File: docs/csharp/classes.md [hunk 15]
```diff
@@ -4,24 +4,78 @@ description: Learn about the class types and how you create them
...
+  
+[!code-csharp[Inheritance](../../samples/snippets/csharp/concepts/classes/inheritance.cs)]  
+  
+When a class declares a base class, it inherits all the members of the base class except the constructors.  
+  
+Unlike C++, a class in C# can only directly inherit from one base class. However, because a base class may itself inherit from another class, a class may indirectly inherit multiple base classes. Furthermore, a class can directly implement more than one interface. For more information, see [Interfaces](interfaces.md).  
+  
+A class can be declared [abstract](https://msdn.microsoft.com/en-us/library/sf985hc5.aspx). An abstract class contains abstract methods that have a signature definition but no implementation. Abstract classes cannot be instantiated. They can only be used through derived classes that implement the abstract methods. By constrast, a [sealed](https://msdn.microsoft.com/en-us/library/88c54tsw.aspx) class does not allow other classes to derive from it. For more information, see [Abstract and sealed classes and class members](https://msdn.microsoft.com/en-us/library/ms173150.aspx).
```
Comment (mairaw) [inline docs/csharp/classes.md]: typo: constrast

File: docs/csharp/structs.md [hunk 16]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+  
+The built-in numeric types are structs, and they have properties and methods that you can access:  
+  
+[!code-csharp[Static Method](../../samples/snippets/csharp/concepts/structs/static-method.cs)]
+  
+But you declare and assign values to them as if they were simple non-aggregate types:  
+  
+[!code-csharp[Assign Values](../../samples/snippets/csharp/concepts/structs/assign-value.cs)] 
+  
+Value types are *sealed*, which means, for example, that you cannot derive a type from @System.Int32, and you cannot define a struct to inherit from any user-defined class or struct because a struct can only inherit from @System.ValueType. However, a struct can implement one or more interfaces. You can cast a struct type to an interface type; this causes a *boxing* operation to wrap the struct inside a reference type object on the managed heap. Boxing operations occur when you pass a value type to a method that takes a @System.Object as an input parameter. For more information, see [Boxing and Unboxing](https://msdn.microsoft.com/en-us/library/yz2be5wk.aspx).
```
Comment (mairaw) [inline docs/csharp/structs.md]: a Object -> an Object

File: docs/csharp/structs.md [hunk 17]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+  
+You use the [struct](https://msdn.microsoft.com/en-us/library/ah19swz4.aspx) keyword to create your own custom value types. Typically, a struct is used as a container for a small set of related variables, as shown in the following example:  
+  
+[!code-csharp[Struct Keyword](../../samples/snippets/csharp/concepts/structs/struct-keyword.cs)]  
+  
+For more information about value types in the .NET Framework, see [Common Type System](../standard/common-type-system.md).  
+    
+Structs share most of the same syntax as classes, although structs are more limited than classes:  
+  
+-   Within a struct declaration, fields cannot be initialized unless they are declared as const or static.
```
Comment (mairaw) [inline docs/csharp/structs.md]: should const and static be formatted as inline code?

File: docs/csharp/structs.md [hunk 18]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+  
+-   Within a struct declaration, fields cannot be initialized unless they are declared as const or static.  
+  
+-   A struct cannot declare a default constructor (a constructor without parameters) or a destructor.  
+  
+-   Structs are copied on assignment. When a struct is assigned to a new variable, all the data is copied, and any modification to the new copy does not change the data for the original copy. This is important to remember when working with collections of value types such as Dictionary<string, myStruct>.  
+  
+-   Structs are value types and classes are reference types.  
+  
+-   Unlike classes, structs can be instantiated without using a **new** operator.
```
Comment (mairaw) [inline docs/csharp/structs.md]: keywords should be styled as inline code, not bold

File: docs/csharp/structs.md [hunk 19]
```diff
@@ -12,16 +12,64 @@ ms.devlang: csharp
...
+  
+-   Structs are copied on assignment. When a struct is assigned to a new variable, all the data is copied, and any modification to the new copy does not change the data for the original copy. This is important to remember when working with collections of value types such as Dictionary<string, myStruct>.  
+  
+-   Structs are value types and classes are reference types.  
+  
+-   Unlike classes, structs can be instantiated without using a **new** operator.  
+  
+-   Structs can declare constructors that have parameters.  
+  
+-   A struct cannot inherit from another struct or class, and it cannot be the base of a class. All structs inherit directly from **System.ValueType**, which inherits from **System.Object**.
```
Comment (mairaw) [inline docs/csharp/structs.md]: same here. either link to the API in this case or style as inline code.

File: samples/snippets/csharp/concepts/classes/class-example.cs [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+}
+class TestPerson
+{
+    static void Main()
+    {
+        // Call the constructor that has no parameters.
+        Person person1 = new Person();
+        Console.WriteLine(person1.name);
+
+        person1.SetName("John Smith");
```
Comment (mairaw) [inline samples/snippets/csharp/concepts/classes/class-example.cs]: there's a process to use names in content. are you aware of the process?

**Review by mairaw — COMMENTED** [review]
Left a few comments for you to address

File: samples/snippets/csharp/concepts/structs/literals.cs (hunk 10 above)
Comment (stevehoag) [inline samples/snippets/csharp/concepts/structs/literals.cs]: Fixed

File: docs/csharp/basic-types.md (hunk 13 above)
Comment (stevehoag) [inline docs/csharp/basic-types.md]: Fixed.

File: docs/csharp/structs.md (hunk 17 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

File: docs/csharp/structs.md (hunk 16 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

File: docs/csharp/classes.md (hunk 15 above)
Comment (stevehoag) [inline docs/csharp/classes.md]: Fixed.

File: docs/csharp/classes.md (hunk 14 above)
Comment (stevehoag) [inline docs/csharp/classes.md]: Fixed.

File: docs/csharp/structs.md (hunk 18 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed. Btw, this and the next one are artifacts from the CAPS markdown conversion - they were styled as bold in the MSDN source.

File: samples/snippets/csharp/concepts/classes/class-example.cs (hunk 20 above)
Comment (stevehoag) [inline samples/snippets/csharp/concepts/classes/class-example.cs]: Yes. This snippet came directly from the MSDN docs.

File: docs/csharp/structs.md (hunk 19 above)
Comment (stevehoag) [inline docs/csharp/structs.md]: Fixed.

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some changed files links of this time:
- Docset 1 - **core-docs**  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134)  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134)  

**Status**: Succeeded.  
[Open Publishing Report](https://opbuildstorageprod.blob.core.windows.net/report/2016%5C10%5C12%5C2a032abb-a76c-bfc8-4231-71ab612989a2%5CPullRequest%5C201610121928084208-1134%5Cworkflow_report.html).

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some changed files links of this time:
- Docset 1 - **core-docs**  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134)  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134)  

**Status**: Succeeded.  
[Open Publishing Report](https://opbuildstorageprod.blob.core.windows.net/report/2016%5C10%5C12%5C2a032abb-a76c-bfc8-4231-71ab612989a2%5CPullRequest%5C201610122017085588-1134%5Cworkflow_report.html).

**Comment by qinezh** [comment]
Open Publishing Build Service: The pull request content has been published and here are some changed files links of this time:
- Docset 1 - **core-docs**  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/basic-types?branch=pr-en-us-1134)  
  [https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134](https://review.docs.microsoft.com/en-us/dotnet/articles/csharp/classes?branch=pr-en-us-1134)  

**Status**: Succeeded.  
[Open Publishing Report](https://opbuildstorageprod.blob.core.windows.net/report/2016%5C10%5C12%5C2a032abb-a76c-bfc8-4231-71ab612989a2%5CPullRequest%5C201610122034438333-1134%5Cworkflow_report.html).

**Comment by stevehoag** [comment]
@BillWagner @mairaw  I've addressed all comments and updated the TOC - should be good to go now.

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
LGTM.  :shipit:

**Pull request merged**